spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/actions/Hive2IcebergAction.java

```diff
@@ -0,0 +1,445 @@
+      return latestVersion;
+    }
+  }
+
+  /**
+   * Step 4: Perform path rewrite and update the metadata of the snapshot table into the original
+   * table.
+   *
+   * @param snapShotTable The snapshot table.
+   * @param sourceTable The original table (Hive table).
+   * @return An object containing the operation status and the latest version of the path rewrite.
+   *     If the path rewrite is successful, it returns true with the corresponding version
+   *     information; if it fails, it returns false.
+   */
+  private OperationPathRewriteResult handlePathRewrite(
+      SparkTable snapShotTable, CatalogTable sourceTable) {
+    try {
+      String sourcePrefix = snapShotTable.table().location();
+      String targetPrefix = CatalogUtils.URIToString(sourceTable.storage().locationUri().get());
+      String stagingLocation = targetPrefix;
+
+      RewriteTablePathSparkAction rewriteAction =
+          SparkActions.get().rewriteTablePath(snapShotTable.table());
+      if (stagingLocation != null) {
+        rewriteAction.stagingLocation(stagingLocation);
+      }
+      rewriteAction.hiveMigrate(true);
+      RewriteTablePath.Result rewrite =
+          rewriteAction.rewriteLocationPrefix(sourcePrefix, targetPrefix).execute();
+      return new OperationPathRewriteResult(true, rewrite.latestVersion());
+    } catch (Exception e) {
+      LOG.error("Path rewrite failed", e);
+      return new OperationPathRewriteResult(false);
+    }
+  }
+
+  /**
+   * Step5. Metadata migration step.
+   *
+   * @param targetPrefix The prefix of the target table's path, used to generate the location for
+   *     the target metadata file.
+   * @param latestVersion The latest version of the target table, which is used to set the metadata
+   *     file's location.
+   * @param sourceIdentifier The identifier of the source table, used to identify and update the
+   *     source table's metadata.
+   * @return Returns true if the metadata migration is successful; otherwise, returns false.
+   */
+  private boolean handleMetadataMigration(
+      String targetPrefix, String latestVersion, Identifier sourceIdentifier) {
+    try {
+      Map<String, String> properties = Maps.newHashMap();
+      properties.put(CatalogProperties.CATALOG_IMPL, HiveCatalog.class.getName());
+
+      HiveCatalog hiveCatalog = new HiveCatalog();
+      hiveCatalog.setConf(spark().sessionState().newHadoopConf());
+      hiveCatalog.initialize("hive", properties);
+
+      String targetMetadataFileLocation = targetPrefix + "/metadata/" + latestVersion;
+      TableCatalog targetSourceCatalog = checkTargetCatalog(sourceCatalog);
+      targetSourceCatalog.alterTable(
+          sourceIdentifier,
+          TableChange.setProperty("write.parquet.compression-codec", "zstd"),
+          TableChange.setProperty("metadata_location", targetMetadataFileLocation),
+          TableChange.setProperty("table_type", "ICEBERG"),
+          TableChange.setProperty("provide", "iceberg"));
```

Comment:
provider?

Reply:
Thank you for paying attention to this PR. I will continue to improve it.